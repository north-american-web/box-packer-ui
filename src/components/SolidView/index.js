import PropTypes from "prop-types";
import React from 'react';

const SolidView = ({ solid, fallbackDescription }) => {
    const description = solid.description ? solid.description : fallbackDescription;
    return (
        <>
            <span className="solid-view__description">{description}</span>
            <span className="solid-view__dimensions">({solid.width}x{solid.length}x{solid.height})</span>
            {solid.hasOwnProperty('contents') && (
                <ul className="solid-preview__children">
                    {solid.contents.map( (child, index) => (
                        <li key={index}>
                            <SolidView fallbackDescription="Item" solid={child} />
                        </li>
                    ))}
                </ul>
            )}
        </>
    )
};

SolidView.propTypes = {
    solid: PropTypes.object.isRequired,
    fallbackDescription: PropTypes.string.isRequired,
    children: PropTypes.array
};

export default SolidView;