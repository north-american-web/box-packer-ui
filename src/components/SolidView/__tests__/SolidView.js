import React from 'react';
import { render, cleanup } from '@testing-library/react';
import 'jest-dom/extend-expect';
import SolidView from '../index';

afterEach(cleanup);

describe('<SolidView/>', () => {
    it('renders', () => {
        const fakeSolid = {
            description: 'Fake item',
            width: 2,
            length: 3,
            height: 4
        };
        const { getByText, asFragment } = render(<SolidView fallbackDescription='Fallback' solid={fakeSolid} />);
        expect(asFragment()).toMatchSnapshot();
        expect(getByText('Fake item')).toHaveClass('solid-view__description');
        expect(getByText('(2x3x4)')).toHaveClass('solid-view__dimensions');
    });

    it('falls back to a default title', () => {
        const fakeSolid = {
            width: 2,
            length: 3,
            height: 4
        };
        const { getByText } = render(<SolidView fallbackDescription='Fallback' solid={fakeSolid} />);
        expect(getByText('Fallback')).toHaveClass('solid-view__description');
    });

    it('renders children when appropriate', () => {
        const fakeSolid = {
            description: 'Fake item',
            width: 2,
            length: 3,
            height: 4,
            contents: [
                {width: 1, length: 1, height: 1},
            ]
        };

        const {getByText, asFragment} = render(<SolidView fallbackDescription='Fallback' solid={fakeSolid}/>);
        expect(asFragment()).toMatchSnapshot();
        expect(getByText('(1x1x1)')).toHaveClass('solid-view__dimensions');
    })
});