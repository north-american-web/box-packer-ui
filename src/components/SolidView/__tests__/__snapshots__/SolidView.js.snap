// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SolidView/> renders 1`] = `
<DocumentFragment>
  <span
    class="solid-view__description"
  >
    Fake item
  </span>
  <span
    class="solid-view__dimensions"
  >
    (2x3x4)
  </span>
</DocumentFragment>
`;

exports[`<SolidView/> renders children when appropriate 1`] = `
<DocumentFragment>
  <span
    class="solid-view__description"
  >
    Fake item
  </span>
  <span
    class="solid-view__dimensions"
  >
    (2x3x4)
  </span>
  <ul
    class="solid-preview__children"
  >
    <li>
      <span
        class="solid-view__description"
      >
        Item
      </span>
      <span
        class="solid-view__dimensions"
      >
        (1x1x1)
      </span>
    </li>
  </ul>
</DocumentFragment>
`;
